veps/sig-storage/dic-on-heterogeneous-cluster/dic-on-heterogeneous-cluster.md

```diff
@@ -0,0 +1,526 @@
+`spec.workloads.nodePlacement` field in the `HyperConverged` CR is defined, HCO will use this field to determine the
+workload nodes instead.
+
+### Reconciling the `SSP` Custom Resource
+When reconciling the `SSP` CR, HCO will:
+
+1. Populate the `spec.cluster.workloadArchitectures` filed in the `SSP` CR with the architectures of the workload nodes
+   currently present in the cluster.
+2. Add the `ssp.kubevirt.io/dict.architectures` annotation to each **predefined** `DataImportCronTemplate` object in the
+ `SSP` CR's `spec.commonTemplates.dataImportCronTemplates` field. This annotation will include a comma-separated list
+ of the supported architectures.
+   > **Note**: For the **custom** `DataImportCronTemplate` objects, HCO will not add this annotation. Users are
+   > responsible for adding the annotation to their custom `DataImportCronTemplate` objects, based on the architectures
+   > supported by the specific image.
+
+For example:
+
+```yaml
+apiVersion: ssp.kubevirt.io/v1beta3
+kind: SSP
+spec:
+  cluster:
+    workloadArchitectures:
+      - amd64
+      - arm64
+      - s390x
+  commonTemplates:
+    dataImportCronTemplates:
+      - metadata:
+          name: "centos-stream9-image-cron"
+          annotations:
+            ssp.kubevirt.io/dict.architectures: "arm64,amd64,s390x"
+    ...
+```
+
+### Reconciling the `DataImportCron` Custom Resources
+
+Upon reconciling the SSP CR, for each `DataImportCronTemplate` object, the SSP operator shall do the following:
+
+1. determine the required architectures. The required architectures are the intersection of
+   the values listed in the `spec.cluster.workloadArchitectures` field, and the
+   `ssp.kubevirt.io/dict.architectures` annotation. For example, if the
+   `ssp.kubevirt.io/dict.architectures` annotation is `"arm64,amd64"` and the
+   `spec.cluster.workloadArchitectures` is `["arm64", "s390x"]`, the required architectures are `["arm64"]`.
+
+   > **Note**: for predefined `DataImportCronTemplate` objects, the `ssp.kubevirt.io/dict.architectures`
+   annotation is expected to be the same as the architectures listed in the `spec.cluster.workloadArchitectures` field.
+   The reason for this behavior is for custom `DataImportCronTemplate`, where the user may want to set the architectures
+   to match the image manifest on one hand, and to avoid creating `DataImportCron` CRs for architectures that are not
+   existing in the cluster.
+2. For each required architecture, the SSP operator shall create or update a `DataImportCron` CR. For each one of these
+   CRs, the SSP operator shall:
+    1. Set the `spec.template.spec.source.registry.platform.architecture` field in the `DataImportCron` CR, to the
+       architecture of the image. This field is used by the cdi-importer to pull the desired image from the registry.
+    2. Add the `template.kubevirt.io/architecture` label with the required architecture.
```

Comment:
why both are needed and not only in the DV template? we should consider whether annotation/field should be added there.

Reply:
Because it makes all the underline CRs, mostly the DataSource, to be searchable by arch and image name (w/o the arch suffix). It is useful for SSP to set the DataSource name in the template, and for the UI to build a proper display of templates.

Reply:
Not sure what duplication you refer to.

Anyway, labels are the straight forward way to filter resources in k8s.

Reply:
Specifying the same architecture & datasource values both in the DataImportCron spec and in its labels is duplication, isn't it?

Reply:
I don't think so. labels are labels and are used for filtering. Spec fields are for the logic. And there are many examples in K8s for similar duplication. 

What is the concern?

Reply:
Ok, got your point.